### DISCUSSION TITLE
Implementation instruction

### DISCUSSION THREAD

**Discussion by Chudinnio** [Set-OutlookSignatures/Set-OutlookSignatures#48] [General]
Hi Markus,
Your solution looks amazing. I would like to implement it in my organization. Is there maybe some kind of manual/guide how to implement Your solution? In file that are in zip i only saw technical requirements.

**Comment by GruberMarkus** [comment]
Hi Chudinnio,

thanks for your nice words.

Let me refer you to the README of Set-OutlookSignatures. While the README is focused rather on technical information and FAQs, and not on organizational aspects, chapter 16.13 will probably help you find what you are looking for:
>16.13 "What is the recommended approach for implementing the software?"
>There is certainly no definitive generic recommendation, but the file '.\docs\Implementation approach.html' should be a good starting point.
>
>The content is based on real life experiences implementing the script in a multi-client environment with a five-digit number of mailboxes.
>
>It contains proven procedures and recommendations for product managers, architects, operations managers, account managers and e-mail and client administrators. It is suited for service providers as well as for clients.
>
>It covers several general overview topics, administration, support, training across the whole lifecycle from counselling to tests, pilot operation and rollout up to daily business.
>
>The document is available in English and German language.
>

**Comment by GruberMarkus** [comment]
P.S.: Should you be looking for consulting, please visit any of my repositories and click the sponsor button, click [here](https://github.com/sponsors/GruberMarkus), or contact me for consulting and project work.

**Comment by GruberMarkus** [comment]
Please let me know if '.\docs\Implementation approach.html' is what you are looking for, and if this issue can marked as solved.

**Comment by donohueb** [comment]
Hi Markus,
this may seem obvious but what file does one actually edit for the variables?
Do we edit Set-OutlookSignatures.ps1 or some other file to put our paths in?
Looks awesome and I want to try it. Just don't know what file to edit to get started.
Thanks

**Reply by GruberMarkus** [reply]
Hi @donohueb,

no, you never edit Set-OutlookSignatures.ps1.

There are basically three configuration files:
- The one used for the connection to Graph. This file is rarely modified, as it is filled with default values that are valid for most environment.
- The one defining replacement variables. This file is modified in more advanced scenarios, and should not be interesing for you right now.
- The one defining which templates are to be used. This is the most modified configuration file.


My recommendation is to start with the README file, which is the document presented right at https://github.com/GruberMarkus/Set-OutlookSignatures. Read the following chapters in order:
- Features: Gives you an overview of what Set-OutlookSignatures can do for you
- Chapter 1, Requirements: Describes the very basis requirements that need to be prepared to run Set-OutlookSignatures in your environment.
- Chapter 2, Parameters: Describes how the behavior of Set-OutlookSignatures can be modified. This not only gives you a deeper understanding of the features, but also answers your question which configuration files are available. The configuration files themselves contain more specific information on how to use them.


Set-OutlookSignatures is very well documented, which inevitably brings with it a lot of content. If you are looking for someone with experience who can quickly train you and assist with evaluation, planning, implementation and ongoing operations, please let me know - I can help with professional services.

**Comment by donohueb** [comment]
Hi Markus,
Ok thank-you. Thanks for responding.
I’m actually working on it now. Trying to get it setup for a test user first. If I get stuck we’ll look at professional services from you.
I was thinking you could put a “Real Beginners” section in the notes as I was reading everywhere but could not figure out which files to edit. So the points you explained to me are good points that should go into your readme file or under one of the chapters.
Anyway I’ll let you know how I go.
Thanks
Ben

**Reply by GruberMarkus** [reply]
You are welcome.

I like the idea of a quick start guide. v4.0.0, whose release is not far in the future, will include such a guide as the first chapter in the README file.

**Reply by GruberMarkus** [reply]
@donohueb ,

what do you think about the following draft of a quick start guide?

# Quick Start Guide
If you already use Set-OutlookSignatures and plan to update to a newer version, start with the CHANGELOG document.

If you are new to Set-OutlookSignatures, start with the README file, which is the document presented right at https://github.com/GruberMarkus/Set-OutlookSignatures - and which you are obviously reading right now.

Read the following chapters in order:
1. Features: Gives you an overview of what Set-OutlookSignatures can do for you.<br>This is the chapter right before this Quick Start Guide.
2. Chapter 1 (Requirements): Describes the very basic requirements that need to be prepared to run Set-OutlookSignatures in your environment. 
3. Chapter 2 (Parameters): Describes how the behavior of Set-OutlookSignatures can be modified.<br>This gives you a deeper understanding of the features, but also answers how you can change the behavior of Set-OutlookSignatures.

You now have a good theoretical overview.  
If you want to know more before you start with the practical implementation, just read through the rest of the README file.

To start with the practical implementation:
1. Log on to a Windows client with a test user having an Exchange mailbox, and configure Outlook
2. Download Set-OutlookSignatures and extract the archive to a local folder
3. Unblock at least 'Set-OutlookSignatures.ps1', you can use the PowerShell commandlet 'Unblock-File' for this.
4. If you use AppLocker or a comparable solution, you may need to digitally sign the PowerShell 'Set-OutlokSignatures.ps1'.  
It is usually not necessary to sign the variable replacement configuration files, e. g. '.\config\default replacement variables.ps1'.  
There are locked down environments, where all files matching the patterns `'*.ps*1'` and `'*.dll'` need to be digitially signed with a trusted certificate. 
5. Now it is time to run Set-OutlookSignatures for the first time
   - Mailbox is in Exchange on-prem and the logged-in user has access to the on-prem Active Directory: Run Set-OutlookSignatures.ps1 in PowerShell.
   - Mailbox is in Exchange Online and/or the logged-in user has no access to the on-prem Active Directory and/or your environment is cloud only: You need to register an Azure AD application first, because Set-OutlookSignatures needs permissions to access the Graph API.
     1. An Azure AD administrator needs to register Set-OutlookSignatures as app and provide admin consent for the required permissions. See '.\config\default graph config.ps1' for details.
     2. Run Set-OutlookSignatures
        - Cloud only, or hybrid without synced Exchange attributes (mail, legacyExchangeDN, msExchRecipientTypeDetails, msExchMailboxGuid, proxyAddresses): '`Set-OutlookSignatures.ps1 -GraphOnly true`'
        - Hybrid with synced Exchange attributes (mail, legacyExchangeDN, msExchRecipientTypeDetails, msExchMailboxGuid, proxyAddresses): '`Set-OutlookSignatures.ps1`'

Set-OutlookSignatures now runs using default settings and sample templates.

Watch the output for errors and warnings, displayed in red or yellow when run in the PowerShell console.
- If there are errors or warnings, read the messages carefully as they often contain hints on how to resolve the issue.  
Check if the README file contains a hint.  
Check if someone has already reported the problem as and issue on [GitHub](https://github.com/GruberMarkus/Set-OutlookSignatures/issues?q=is%3Aissue), and create a new one if you can't find any hint on how to solve it.
- If there are no errors, switch to Outlook and have a look at the newly created signatures.

When everything runs fine with default settings, it is time to start custominzing the script behavior to your needs:
- Create a folder with your own template files and signature configuration file.
  - Start with DOCX templates. See `'Should I use .docx or .htm as file format for templates?'` in this document for details.
  - See the following chapters in this document for instructions
    - Signature and OOF file format
    - Signature template file naming
    - Template tags and ini files
  - Make sure to pass the parameters `'SignatureTemplatePath'`, `'SignatureIniPath'`, `'OOFTemplatePath'` and `'OOFInipath'` to Set-OutlookSignatures
- Adopt other parameters you may find useful, or start experimenting with simulation mode.

It is strongly recommended to not change any Set-OutlookSignature files and keep them as they are. If you consequently work with script parameters and keep customized configuration files in a separate folder, upgrading to a new version is basically just a file copy operation (drop-in replacement).

Regarding configuration files: Besides the template configuration files for signatures and OOF messages, there are the Graph configuration file and the replacement variable configuration file.  
It is rarely needed to change the configuration within these files. The files themselves contain specific information on how to use them.

**Comment by donohueb** [comment]
Hi Markus,
I think the Features should he chapter 1.
Chapter 2. Requirements
Chapter 3. Quick Start Guide
Chapter 4. etc...

Need a way to pause the script before exiting. I’m getting an error and the screen flashes off before I can get a screenshot. How to pause the script?
Need the simpliest way to run it all from c:\set-outlooksignatures in the quickstart.
So the absolute minimum to get started and then can get complex from there.

Thanks

**Reply by GruberMarkus** [reply]
I will think about adding more instructions on how to run the script.

Regarding the script error: It sounds like you are running into a basic thing with PowerShell I guess you execute the .ps1 in File Explorer via right-click. Instead:
- Run powershell.exe
- & 'c:\signature\set-outlooksignatures.ps1' # adopt path to your actual location

**Comment by donohueb** [comment]
Ok thanks I got it working.

Perhaps have a command line script for a logged in user as in c:\users\test\downloads\    and everything running from in here.
Please see attached new error. Strange.
Does the script have to run every time a user logs in or just once ever?
What is the default file that below is coming from? This is a test user.

***@***.***

Once I get it up and running I’ll be happy to write up a real beginners guide for you if you like.
I’ll keep it short and a minimum setup to get it up and running.
Thanks

**Reply by GruberMarkus** [reply]
Hi Ben,

there is no attachment to review. It is usually better to anwer in GitHub directly, and not via e-mail.

It is up to you how often the script runs - at logon, every two hours, only when users start it themselves, etc.

**Comment by donohueb** [comment]
Ok I’ll post errors on github.
Thanks for the responses.
